source/v1/I2CDetail.cpp

```diff
@@ -0,0 +1,235 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2016 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed-drivers/v1/I2CDetail.hpp"
+
+#if DEVICE_I2C && DEVICE_I2C_ASYNCH
+
+#include "mbed-drivers/v1/I2C.hpp"
+#include "core-util/CriticalSectionLock.h"
+#include "minar/minar.h"
+
+namespace mbed_drivers {
+using namespace mbed;
+namespace v1 {
+namespace detail {
+// TODO: Increase the size of I2COwners to accept Resource Mangers for non-onchip I2C masters
+I2CResourceManager * I2COwners[MODULES_SIZE_I2C] = {nullptr};
+
+int I2CResourceManager::post_transaction(I2CTransaction *t) {
+    int rc;
+    if ((rc = validate_transaction(t))) {
+        return rc;
+    }
+    // Optimization NOTE: this could be lock-free with atomic_cas.
+    mbed::util::CriticalSectionLock lock;
+    // Find the tail of the queue
+    I2CTransaction * tx = (I2CTransaction *)TransactionQueue;
+    if (!tx) {
+        TransactionQueue = t;
+        power_up();
+        start_transaction();
+        return 0;
+    }
+    while (tx->next) {
+        tx = tx->next;
+    }
+    tx->next = t;
+    return 0;
+}
+
+void I2CResourceManager::process_event(int event) {
+    I2CTransaction * t;
+    CORE_UTIL_ASSERT(TransactionQueue != nullptr);
+    if (!TransactionQueue) {
+        return;
+    }
+    // Get the current item in the transaction queue
+    t = (I2CTransaction *) TransactionQueue;
```

Comment:
Oddly enough because the C++ cast mechanism isn't adequate to strip a volatile qualifier.